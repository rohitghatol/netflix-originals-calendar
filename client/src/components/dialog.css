.link {
    text-decoration: none;
}

.link:hover {
    text-decoration: none;
    color: black;
}

.cover {
    position: fixed;
    left: 0px;
    top: 0px;
    width: 100%;
    height: 100%;
    opacity: 20%;
    z-index: 1000;
    background: #aaaaaa;
    opacity: 0.2;
}
.dialog {
    display:grid;
    position: fixed;
    left: 25%;
    top: 25%;
    width: 50%;
    height: 50%;
    z-index: 1001;
    border: 1px solid darkslategray;
    border-radius: 4px;
    box-shadow: 10px 10px 5px 0px rgba(0,0,0,0.75);
    grid-template-rows: 40px 1fr;
}

.header {
    display:grid;
    grid-template-columns: 1fr 30px;
    background: #444444;
    color: white;
    align: center;
    padding:10px;
}
/*FIXME - Pending handle scroll horizontal as well as vertical*/
.content {
    background: #eee;
    padding:10px;
    overflow: auto;
}

.button {
    cursor: pointer;
}
.button:hover {
    color: red;
}

.text {
    white-space: nowrap;
    overflow: hidden;
    font-size: 10pt;
}
