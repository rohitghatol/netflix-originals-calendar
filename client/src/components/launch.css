.truncate {
    width: 140px; /* Use parent CSS Grid to avoid setting fixed width for ecllipsis */
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: 10pt;
}
