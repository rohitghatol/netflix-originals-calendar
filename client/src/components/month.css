.month {
    display: grid;
    grid-template-rows: 1fr 1fr 10fr;
    height: 100%;
}
.navigation {
    display: grid;
    grid-template-columns: 1fr 32px 200px 32px 1fr
}
.action {
    font-size: 24pt;
    user-select: none;
    cursor: pointer;
    padding:10px;
    color: #aaa;
}

.action:active {
    font-weight: bolder;
}

.action:hover {
     color: #444;
}

.title:hover {
    color: #444;
}

.title,.action {
    display:grid;
    align-content: center;
    justify-self: center;
}

.weeks {
    display:grid;
    grid-template-columns: repeat(7,1fr);
    grid-template-rows: 1fr;
    border-right: 1px solid lightgray;
}

.weekDay {
    display:grid;
    padding:10px;
    align-content: center;
    border-top: 1px solid lightgray;
    border-left: 1px solid lightgray;
}
.weekDay:hover {
    background: lightblue;
}

.days {
    display:grid;
    grid-template-columns: repeat(7,1fr);
    grid-template-rows: repeat(6,1fr);
    border-right: 1px solid lightgray;
    border-bottom: 1px solid lightgray;
 }

.day {
    padding:10px;
    border-top: 1px solid lightgray;
    border-left: 1px solid lightgray;
}

.day:hover {
    background: #eee;
}

