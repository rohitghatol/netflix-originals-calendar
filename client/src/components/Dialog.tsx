import React, {useState} from 'react';
import {ILaunch} from '../types';
import './dialog.css';

export interface IProps {
  day: number;
  launches: ILaunch[];
}

export const Dialog = (props: IProps) => {
  const [open, setOpen] = useState(false);
  const {launches, day} = props;
  const onClick = () => {
    setOpen(!open);
  };
  return (
    <div>
      <a className='link' href='#' onClick={onClick}>...</a>
      <div className='cover' style={{display: open ? 'grid' : 'none'}} onClick={onClick}/>
      <div className='dialog' style={{display: open ? 'grid' : 'none'}}>
        <div className='header'>
          <div>Launches - {day}</div>
          <div className='button' onClick={() => {setOpen(false); }}>x</div>
        </div>
        <div className='content'>
          <div>
            {
              (launches || []).map( (launch: ILaunch) => ( <div key={launch.id} className='text'> {launch.title} </div> ))
            }
          </div>
        </div>
      </div>
    </div>
  );
};
