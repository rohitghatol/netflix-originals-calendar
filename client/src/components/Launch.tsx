import React from 'react';
import {ILaunch} from '../types';

import './launch.css';
import {Dialog} from './Dialog';


export interface IProps {
  day: number;
  launches: ILaunch[];
}

export const Launch = (props: IProps) => {
  const {launches, day} = props;
  const visibleLaunches = launches.slice(0, 2);

  return (
    <div>
    {
      (visibleLaunches || []).map( (launch: ILaunch) => ( <div key={launch.id} className='truncate'> {launch.title} </div> ))
    }
    <Dialog day={day} launches={launches}/>
    </div>

  );
};
