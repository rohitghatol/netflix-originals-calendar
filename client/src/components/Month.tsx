import React from 'react';
import _ from 'lodash';
import {calendar} from '../service';
import {Launch} from './Launch';


import './month.css';


export interface IProps {
  year: number;
  month: number;
  launches: any; // {[key: string]: ILaunch[]};
  prev: Function;
  next: Function;
}

export const Month = (props: IProps) => {
  const {month, year, prev, next, launches} = props;
  const weekdays = calendar.getWeekDays();
  const startIndex = calendar.firstDayInMonth(month, year);
  const totalDays = calendar.daysInMonth(month, year);
  const days = new Array(42).fill(null);
  let day = 1;
  for (let index = startIndex; index < totalDays + startIndex; index++) {
    days[index] = day++;
  }
  return (
    <div className='month'>

      <div className='navigation'>
        <div/>
        <div className='action' onClick={() => {
          prev();
        }}>&larr;</div>
        <div className='title'>{calendar.getMonth(month)} {year}</div>
        <div className='action' onClick={() => {
          next();
        }}>&rarr;</div>
        <div/>
      </div>

      <div className='weeks'>
        {weekdays.map(name => (<div key={name} className='weekDay'>{name}</div>))}
      </div>
      <div className='days'>
        {days.map((day, index) => {
            if (day) {
              const key = `${year}-${_.padStart(month.toString(), 2, '0')}-${_.padStart(day.toString(), 2, '0')}`;
              const launchesForDay = launches[key];
              return (
                <div key={index} className='day'>
                  <div>
                    {day}
                  </div>
                  {launchesForDay ?
                    <Launch day={day} launches={launchesForDay}/>
                    : null}
                </div>);
            }
            else {
              return (
                <div key={index} className='day'>
                    <div>
                      {day}
                    </div>
                </div>
              );
            }
        })}

      </div>
    </div>
  );
};
