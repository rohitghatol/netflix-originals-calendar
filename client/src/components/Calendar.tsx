import React, {useState, useEffect} from 'react';
import {withRouter} from 'react-router-dom';
import {ILaunch} from '../types';
import {calendar} from '../service';
import {Month} from './Month';

export interface IProps {
  match: any;
  history: any;
  location: any;
}

export interface IState {
  month: number;
  year: number;
}

const Calendar = (props: IProps) => {
  const [currentYear, setCurrentYear] = useState(calendar.currentMonth().year);
  const [currentMonth, setCurrentMonth] = useState(calendar.currentMonth().month);
  const [launches, setLaunches] = useState({});
  useEffect(() => {
    if (props.match && props.match.params && props.match.params.year && props.match.params.month) {
      setCurrentMonth(parseInt(props.match.params.month));
      setCurrentYear(parseInt(props.match.params.year));
    } else {
      const {month, year} = calendar.currentMonth();
      setCurrentMonth(month);
      setCurrentYear(year);
    }
  }, [props.match]);
  useEffect(() => {
    // console.log('Fetch Data');
    const fetchData = async () => {
      const response = await fetch(`/api/events`, { method: 'GET' });
      const json = await response.json();
      // FIXME - Since we are not really calling an api like /api/events/year/month
      // we are getting complete data.
      // The cost for filtering this list vs cost to create a map is almost the same
      // Ideally for an /api/events/:year/:month we would get much smaller set
      const launchMap = {};
      json.data.forEach( (event: ILaunch) => {
        const key = event.launch_date.substring(0, 10);
        if (!launchMap[key]) {
          launchMap[key] = [];
        }
        launchMap[key].push(event);
      });
      setLaunches(launchMap);
    };
    fetchData();
  }, [currentMonth, currentYear]);

  const {history} = props;
  const prev = () => {
    const {month, year} = calendar.prevMonth(currentMonth, currentYear);
    history.push(`/${year}/${month}`);
  };

  const next = () => {
    const {month, year} = calendar.nextMonth(currentMonth, currentYear);
    history.push(`/${year}/${month}`);
  };

  return <Month
    year={currentYear}
    month={currentMonth}
    launches={launches}
    next={next}
    prev={prev}
  />;
};


export const CalendarWithRouter = withRouter(Calendar);
