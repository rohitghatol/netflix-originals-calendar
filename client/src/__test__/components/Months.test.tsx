// Month.react.test.js
import React from 'react';
import {Month} from '../../components/Month';
import renderer from 'react-test-renderer';

test('Month renders calendar month', () => {
  const component = renderer.create(
    <Month month={1} year={2019} launches={{}} next={()=>{}} prev={()=>{}}/>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
