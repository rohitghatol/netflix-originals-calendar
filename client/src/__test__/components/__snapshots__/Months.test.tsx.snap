// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Month renders calendar month 1`] = `
<div
  className="month"
>
  <div
    className="navigation"
  >
    <div />
    <div
      className="action"
      onClick={[Function]}
    >
      ←
    </div>
    <div
      className="title"
    >
      January
       
      2019
    </div>
    <div
      className="action"
      onClick={[Function]}
    >
      →
    </div>
    <div />
  </div>
  <div
    className="weeks"
  >
    <div
      className="weekDay"
    >
      Su
    </div>
    <div
      className="weekDay"
    >
      Mo
    </div>
    <div
      className="weekDay"
    >
      Tu
    </div>
    <div
      className="weekDay"
    >
      We
    </div>
    <div
      className="weekDay"
    >
      Th
    </div>
    <div
      className="weekDay"
    >
      Fr
    </div>
    <div
      className="weekDay"
    >
      Sa
    </div>
  </div>
  <div
    className="days"
  >
    <div
      className="day"
    >
      <div />
    </div>
    <div
      className="day"
    >
      <div />
    </div>
    <div
      className="day"
    >
      <div>
        1
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        2
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        3
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        4
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        5
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        6
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        7
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        8
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        9
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        10
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        11
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        12
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        13
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        14
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        15
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        16
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        17
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        18
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        19
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        20
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        21
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        22
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        23
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        24
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        25
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        26
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        27
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        28
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        29
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        30
      </div>
    </div>
    <div
      className="day"
    >
      <div>
        31
      </div>
    </div>
    <div
      className="day"
    >
      <div />
    </div>
    <div
      className="day"
    >
      <div />
    </div>
  </div>
</div>
`;
